python/tank/platform/environment.py

```diff
@@ -878,3 +880,262 @@ def copy_apps(self, src_engine_name, dst_engine_name):
+                    # Add a comment to identify default values in the dumped
+                    # file to help with debugging. The comment shows the
+                    # default value and the path to the manifest where the
+                    # default value is defined.
+                    if hasattr(settings, 'yaml_add_eol_comment'):
+                        settings.yaml_add_eol_comment(
+                            "Differs from %s default value (%s) in manifest %s"
+                                % (engine_name or "",
+                                   schema_default or '""',
+                                   manifest_file or ""
+                                ),
+                            setting_name,
+                            column=80
+                        )
+
+                # identify the type to address any special cases
+                setting_type = setting_schema["type"]
+
+                # remove any legacy "default" hook references
+                if setting_type == "hook" and setting_value == "default":
+                    del settings[setting_name]
+                    modified = True
+
+        return modified
+
+    def _update_settings_full(self, schema, settings, engine_name=None,
+        manifest_file=None, debug=False):
+        """
+        Given a schema and settings, make sure that the settings are fully
+        defined.
+
+        Iterate over the schema key's default values and add any settings
+        that aren't defined by using the default value.
+
+        :param schema: A schema defining types and defaults for settings.
+        :param settings: A dict of settings to sparsify.
+        :param engine_name: The name of the engine currently running if there
+            is one.
+        :param manifest_file: The path to the manifest file if known.
+        :param debug: If True, include debug comments on lines using a default
+            value.
+
+        :returns: bool - True if the settings were modified, False otherwise.
+        """
+
+        modified = False
+
+        # check each key defined in the schema
+        for setting_name in schema.keys():
+
+            # get the default value for the setting
+            setting_schema = schema[setting_name]
+            schema_default = resolve_default_value(
+                setting_schema, engine_name=engine_name)
+
+            if setting_name not in settings:
+                # No value specified in the settings. Just set the default.
+                settings[setting_name] = schema_default
+                modified = True
+
+            if schema_default == settings[setting_name] and debug:
+                # The value of the setting matches the default value in the
+                # manifest. Add a comment to identify default values in the
+                # dumped file to help with debugging. The comment will display
+                # the path to the manifest.
+                if hasattr(settings, 'yaml_add_eol_comment'):
```

Comment:
This works with the new parser which is why it's checking for the attribute here. I have added a note in the usage that the flag only works with the new parser (I noted the core version number it was introduced) so that people shouldn't be surprised if they don't get the comments when they use "--dump-debug"